gpu/gpu.go

```diff
@@ -0,0 +1,134 @@
+//go:build linux || windows
+
+package gpu
+
+/*
+#cgo linux LDFLAGS: -lrt -lpthread -ldl -lstdc++ -lm
+#cgo windows LDFLAGS: -lpthread
+
+#include "gpu_info.h"
+
+*/
+import "C"
+import (
+	"fmt"
+	"log"
+	"sync"
+	"unsafe"
+
+	"github.com/jmorganca/ollama/api"
+)
+
+type handles struct {
+	cuda *C.cuda_handle_t
+	rocm *C.rocm_handle_t
+}
+
+var gpuMutex sync.Mutex
+var gpuHandles *handles = nil
+
+// Note: gpuMutex must already be held
+func initGPUHandles() {
+	// TODO - if the ollama build is CPU only, don't do these checks as they're irrelevant and confusing
+	log.Printf("Detecting GPU type")
+	gpuHandles = &handles{nil, nil}
+	var resp C.cuda_init_resp_t
+	C.cuda_init(&resp)
```

Comment:
It would make things considerably easier IMO if errors were well defined and `*_init` returned an int error. e.g.

```go
var cuda C.cuda_init_resp_t
if retval := C.cuda_init(&cuda); retval == 0 {
  gpuHandles.cuda = &cuda.ch
  return
}

var rocm C.rocm_init_resp_t
if retval := C.rocm_init(&rocm); retval == 0 {
  gpuHandles.rocm = &rocm.rh
  return
}

// cpu
```

Reply:
I'll take a look at tidying up the error handling in a follow up PR.